crates/ra_assists/src/assists/invert_if.rs

```diff
@@ -0,0 +1,89 @@
+use hir::db::HirDatabase;
+use ra_syntax::ast::{self, AstNode};
+use ra_syntax::{TextRange, TextUnit};
+
+use super::apply_demorgan::undo_negation;
+use crate::{Assist, AssistCtx, AssistId};
+
+// Assist: invert_if
+//
+// Apply invert_if
+// This transforms if expressions of the form `if !x {A} else {B}` into `if x {B} else {A}`
+// This also works with `!=`. This assist can only be applied with the cursor
+// on `if`.
+//
+// ```
+// fn main() {
+//     if<|> !y {A} else {B}
+// }
+// ```
+// ->
+// ```
+// fn main() {
+//     if y {B} else {A}
+// }
+// ```
+
+pub(crate) fn invert_if(ctx: AssistCtx<impl HirDatabase>) -> Option<Assist> {
+    let expr = ctx.find_node_at_offset::<ast::IfExpr>()?;
+    let expr_range = expr.syntax().text_range();
+    let if_range = TextRange::offset_len(expr_range.start(), TextUnit::from_usize(2));
+    let cursor_in_range = ctx.frange.range.is_subrange(&if_range);
+    if !cursor_in_range {
+        return None;
+    }
+
+    let cond = expr.condition()?.expr()?.syntax().clone();
+    let then_node = expr.then_branch()?.syntax().clone();
+
+    if let ast::ElseBranch::Block(else_block) = expr.else_branch()? {
+        let flip_cond = undo_negation(cond.clone())?;
```

Comment:
This might be a good opportunity for refactoring: `undo_negation(node: SyntaxNode) -> SyntaxNode` -> `invert_boolean_expression(expr: ast::Expr) -> ast::Expr`

Reply:
How could I create a new `Expr` in ra_assist crate? For example, `1 != 2` flip to `1 == 2` need a new `Expr`.  I don't know how to create a new `SyntaxNode`. Maybe use a `SyntaxTreeBuilder` from `ra_syntax` crate?

Reply:
See `edit` and `make` modules in `ra_syntax`, and various replace and insert functions in the `algo` module.  Note that some editing functions work on an untyped `SyntaxNode` tree, you can get back to typed `ast::FooDef` by casting